File: lib/feature/_attendance_/attendance_analytics/data/attendance_analytics.dart
import 'package:clover/feature/_attendance_/attendance_analytics/data/models/attendance_analytics_models.dart';
import 'package:clover/feature/_attendance_/shared/data/models/attendance_absence.dart';
import 'package:clover/feature/_attendance_/shared/data/models/attendance_punch_record.dart';
import 'package:clover/feature/_attendance_/shared/data/models/attendance_snapshot.dart';
import 'package:clover/feature/_attendance_/shared/data/models/attendance_workplace.dart';

/// Аналитика из snapshot (punches / absences / workers) — без mock-паттернов.
abstract final class AttendanceAnalytics {
  static DateTime dayKey(DateTime value) => DateTime(value.year, value.month, value.day);

  static DateTime get today => dayKey(DateTime.now());

  static String _hhmm(DateTime at) =>
      '${at.hour.toString().padLeft(2, '0')}:${at.minute.toString().padLeft(2, '0')}';

  static AttendanceAnalyticsOverview overview({
    required AttendanceSnapshot snapshot,
    required String workplaceId,
    required DateTime start,
    required DateTime end,
  }) {
    final workplace = snapshot.workplaceById(workplaceId);
    final workers = snapshot.workersFor(workplaceId).where((w) => w.isAccepted).toList(growable: false);
    final built = workers
        .map(
          (w) => worker(
            snapshot: snapshot,
            workplaceId: workplaceId,
            workerId: w.id,
            displayName: w.displayName,
            username: w.username,
            start: start,
            end: end,
            workplace: workplace,
          ),
        )
        .whereType<AttendanceAnalyticsWorker>()
        .toList(growable: false);

    final total = built.fold<int>(0, (sum, w) => sum + w.totalMinutes);
    final avg = built.isEmpty ? 0 : total ~/ built.length;
    final late = built.fold<int>(0, (sum, w) => sum + w.lateDays);
    final missed = built.fold<int>(0, (sum, w) => sum + w.missedDays);
    built.sort((a, b) => b.totalMinutes.compareTo(a.totalMinutes));

    return AttendanceAnalyticsOverview(
      workers: built,
      totalMinutes: total,
      avgMinutesPerWorker: avg,
      lateDaysTotal: late,
      missedDaysTotal: missed,
    );
  }

  static AttendanceAnalyticsWorker? worker({
    required AttendanceSnapshot snapshot,
    required String workplaceId,
    required String workerId,
    required DateTime start,
    required DateTime end,
    String? displayName,
    String? username,
    AttendanceWorkplace? workplace,
  }) {
    final listItem = snapshot.workersFor(workplaceId).where((w) => w.id == workerId).firstOrNull;
    final name = displayName ?? listItem?.displayName ?? _shortId(workerId);
    final nick = username ?? listItem?.username ?? '';
    final wp = workplace ?? snapshot.workplaceById(workplaceId);

    final rangeStart = dayKey(start);
    final rangeEnd = dayKey(end);
    final absences = snapshot.absences.where(
      (a) => a.workplaceId == workplaceId && a.workerId == workerId,
    );
    final punches = snapshot.punchHistory
        .where((p) => p.workplaceId == workplaceId && p.workerId == workerId && !p.cancelled)
        .toList(growable: false);

    final days = <DateTime, AttendanceWorkerDayRecord>{};
    var total = 0;
    var worked = 0;
    var late = 0;
    var missed = 0;

    for (var d = rangeStart; !d.isAfter(rangeEnd); d = d.add(const Duration(days: 1))) {
      final key = dayKey(d);
      final record = _dayRecord(
        key: key,
        punches: punches,
        absences: absences,
        workplace: wp,
        today: today,
      );
      days[key] = record;

      total += record.totalMinutes;
      if (record.status == AttendanceDayStatus.full ||
          record.status == AttendanceDayStatus.late ||
          record.status == AttendanceDayStatus.partial) {
        worked++;
      }
      if (record.status == AttendanceDayStatus.late) late++;
      if (record.status == AttendanceDayStatus.absent || record.status == AttendanceDayStatus.partial) {
        missed++;
      }
    }

    return AttendanceAnalyticsWorker(
      id: workerId,
      displayName: name,
      username: nick,
      totalMinutes: total,
      daysWorked: worked,
      lateDays: late,
      missedDays: missed,
      daysByKey: days,
    );
  }

  static AttendanceWorkerDayRecord _dayRecord({
    required DateTime key,
    required List<AttendancePunchRecord> punches,
    required Iterable<AttendanceAbsenceEntry> absences,
    required AttendanceWorkplace? workplace,
    required DateTime today,
  }) {
    if (key.isAfter(today)) {
      return AttendanceWorkerDayRecord(date: key, status: AttendanceDayStatus.off, totalMinutes: 0);
    }

    if (absences.any((a) => a.covers(key))) {
      return AttendanceWorkerDayRecord(date: key, status: AttendanceDayStatus.excused, totalMinutes: 0);
    }

    final dayPunches = punches.where((p) => dayKey(p.at) == key).toList(growable: false)
      ..sort((a, b) => a.at.compareTo(b.at));

    if (dayPunches.isEmpty) {
      if (key.weekday == DateTime.saturday || key.weekday == DateTime.sunday) {
        return AttendanceWorkerDayRecord(date: key, status: AttendanceDayStatus.off, totalMinutes: 0);
      }
      return AttendanceWorkerDayRecord(date: key, status: AttendanceDayStatus.absent, totalMinutes: 0);
    }

    final labels = dayPunches
        .map(
          (p) => AttendanceDayPunch(
            timeLabel: _hhmm(p.at),
            label: p.type.labelRu,
          ),
        )
        .toList(growable: false);

    DateTime? firstIn;
    DateTime? lastOut;
    for (final p in dayPunches) {
      if (p.type.isClockIn && firstIn == null) firstIn = p.at;
      if (p.type.isClockOut) lastOut = p.at;
    }

    var minutes = 0;
    if (firstIn != null && lastOut != null && !lastOut.isBefore(firstIn)) {
      minutes = lastOut.difference(firstIn).inMinutes;
    }

    final scheduled = workplace?.clockInScheduledTime;
    var status = AttendanceDayStatus.full;
    if (firstIn != null && lastOut == null) {
      status = AttendanceDayStatus.partial;
    } else if (firstIn != null && scheduled != null) {
      final scheduledAt = DateTime(key.year, key.month, key.day, scheduled.hour, scheduled.minute);
      if (firstIn.isAfter(scheduledAt.add(const Duration(minutes: 5)))) {
        status = AttendanceDayStatus.late;
      }
    }

    return AttendanceWorkerDayRecord(
      date: key,
      status: status,
      totalMinutes: minutes,
      punches: labels,
    );
  }

  static String _shortId(String id) => id.length > 8 ? '${id.substring(0, 8)}…' : id;
}

extension _FirstOrNullAnalytics<E> on Iterable<E> {
  E? get firstOrNull {
    final it = iterator;
    if (!it.moveNext()) return null;
    return it.current;
  }
}

File: web/src/features/attendance/lib/attendance-analytics.ts
import {
  dayKey,
  formatDateKey,
  formatMinutes,
  isoWeekday,
  parseDateKey,
  type AttendanceAbsence,
  type AttendanceWorkplace,
} from "@/features/attendance/lib/attendance-model";

export type AnalyticsPunch = {
  id: string;
  profileId: string;
  punchKind: string;
  punchedAt: Date;
  cancelled: boolean;
};

export type DayStatus =
  | "full"
  | "late"
  | "partial"
  | "absent"
  | "excused"
  | "off";

export type WorkerDay = {
  dateKey: string;
  status: DayStatus;
  totalMinutes: number;
  lateMinutes: number;
};

export type AnalyticsWorker = {
  id: string;
  displayName: string;
  username: string;
  totalMinutes: number;
  daysWorked: number;
  lateDays: number;
  missedDays: number;
  days: WorkerDay[];
};

export type AnalyticsOverview = {
  workers: AnalyticsWorker[];
  totalMinutes: number;
  avgMinutesPerWorker: number;
  lateDaysTotal: number;
  missedDaysTotal: number;
};

export function mapOverviewPunch(
  raw: Record<string, unknown>,
): AnalyticsPunch | null {
  const id = String(raw.id ?? "").trim();
  const profileId = String(raw.profile_id ?? "").trim();
  const punchedAt = new Date(String(raw.punched_at ?? ""));
  if (!id || !profileId || Number.isNaN(punchedAt.getTime())) return null;
  return {
    id,
    profileId,
    punchKind: String(raw.punch_kind ?? "clock_in"),
    punchedAt,
    cancelled: raw.cancelled_at != null,
  };
}

function coversAbsence(a: AttendanceAbsence, key: Date): boolean {
  const start = dayKey(parseDateKey(a.startDate));
  const end = dayKey(parseDateKey(a.endDate));
  return key >= start && key <= end;
}

function dayRecord(params: {
  key: Date;
  punches: AnalyticsPunch[];
  absences: AttendanceAbsence[];
  workplace: AttendanceWorkplace;
  today: Date;
}): WorkerDay {
  const { key, punches, absences, workplace, today } = params;
  const dateKey = formatDateKey(key);
  if (key > today) {
    return { dateKey, status: "off", totalMinutes: 0, lateMinutes: 0 };
  }
  if (absences.some((a) => coversAbsence(a, key))) {
    return { dateKey, status: "excused", totalMinutes: 0, lateMinutes: 0 };
  }
  const dayPunches = punches
    .filter((p) => !p.cancelled && formatDateKey(dayKey(p.punchedAt)) === dateKey)
    .sort((a, b) => a.punchedAt.getTime() - b.punchedAt.getTime());

  if (dayPunches.length === 0) {
    const wd = isoWeekday(key);
    if (wd === 6 || wd === 7) {
      return { dateKey, status: "off", totalMinutes: 0, lateMinutes: 0 };
    }
    return { dateKey, status: "absent", totalMinutes: 0, lateMinutes: 0 };
  }

  let firstIn: Date | null = null;
  let lastOut: Date | null = null;
  for (const p of dayPunches) {
    if (p.punchKind === "clock_in" && !firstIn) firstIn = p.punchedAt;
    if (p.punchKind === "clock_out") lastOut = p.punchedAt;
  }

  let minutes = 0;
  if (firstIn && lastOut && lastOut >= firstIn) {
    minutes = Math.round((lastOut.getTime() - firstIn.getTime()) / 60000);
  }

  let status: DayStatus = "full";
  let lateMinutes = 0;
  if (firstIn && !lastOut) {
    status = "partial";
  } else if (firstIn && workplace.clockInScheduled) {
    const [hh, mm] = workplace.clockInScheduled.split(":").map(Number);
    const scheduledAt = new Date(
      key.getFullYear(),
      key.getMonth(),
      key.getDate(),
      hh ?? 0,
      mm ?? 0,
    );
    if (firstIn.getTime() > scheduledAt.getTime() + 5 * 60000) {
      status = "late";
      lateMinutes = Math.round(
        (firstIn.getTime() - scheduledAt.getTime()) / 60000,
      );
    }
  }

  return { dateKey, status, totalMinutes: minutes, lateMinutes };
}

export function buildAnalyticsOverview(params: {
  workplace: AttendanceWorkplace;
  workerIds: string[];
  labels: Map<string, { name: string; username: string }>;
  punches: AnalyticsPunch[];
  absences: AttendanceAbsence[];
  start: string;
  end: string;
}): AnalyticsOverview {
  const start = dayKey(parseDateKey(params.start));
  const end = dayKey(parseDateKey(params.end));
  const today = dayKey(new Date());
  const workers: AnalyticsWorker[] = [];

  for (const id of params.workerIds) {
    const label = params.labels.get(id);
    const punches = params.punches.filter((p) => p.profileId === id);
    const absences = params.absences.filter((a) => a.profileId === id);
    const days: WorkerDay[] = [];
    let total = 0;
    let worked = 0;
    let late = 0;
    let missed = 0;
    for (let d = new Date(start); d <= end; d.setDate(d.getDate() + 1)) {
      const key = dayKey(d);
      const record = dayRecord({
        key,
        punches,
        absences,
        workplace: params.workplace,
        today,
      });
      days.push(record);
      total += record.totalMinutes;
      if (
        record.status === "full" ||
        record.status === "late" ||
        record.status === "partial"
      ) {
        worked += 1;
      }
      if (record.status === "late") late += 1;
      if (record.status === "absent" || record.status === "partial") missed += 1;
    }
    workers.push({
      id,
      displayName: label?.name ?? id.slice(0, 8),
      username: label?.username ?? "",
      totalMinutes: total,
      daysWorked: worked,
      lateDays: late,
      missedDays: missed,
      days,
    });
  }

  workers.sort((a, b) => b.totalMinutes - a.totalMinutes);
  const totalMinutes = workers.reduce((s, w) => s + w.totalMinutes, 0);
  return {
    workers,
    totalMinutes,
    avgMinutesPerWorker:
      workers.length === 0 ? 0 : Math.floor(totalMinutes / workers.length),
    lateDaysTotal: workers.reduce((s, w) => s + w.lateDays, 0),
    missedDaysTotal: workers.reduce((s, w) => s + w.missedDays, 0),
  };
}

export { formatMinutes };
